### ISSUE TITLE
Allow SKIP_INPUT for RNNs

### ISSUE THREAD

**Issue by SeanNaren** [pytorch/pytorch#633]
cuDNN supports skipping the matrix multiplication with the incoming input via the `CUDNN_SKIP_INPUT` flag on the first recurrent layer. This was useful for passing in a pre-batchnormed input to the RNN.

Currently it's hardcoded to linear [here](https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/rnn.py#L157) and I assume this is due to being able to seamlessly move between torch RNNs and cudnn RNNs.

Would I be correct in thinking that just [these](https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/rnn.py#L10-L47) cells need to be updated to take into consideration the flag? If so I'll get to work!

**Comment by adamlerer** [comment]
I wasn't aware this was a common use case. If you would like to add it, you'll need to add support for it to AutogradRNN as well, because we require them to be fully equivalent (so that you can run your RNNs without cudnn or on cpu). Other than that, it sounds good to me.

**Comment by apaszke** [comment]
@adamlerer this would be also helpful to implement that batched input mm for non-cudnn RNNs. We could just apply cells that have no input weights to precomputed values.

**Comment by ngimel** [comment]
A note on cudnn compatibility: up to 5.1.10, even in SKIP_INPUT mode cudnn still required allocating memory for input transformation weights, and IIRC in cudnnGetRNNLinLayerMatrixParams returned non-zero-element descriptors for input transformation matrices. It was fixed for cudnn 5.1.10 and cudnn v6, but I guess it should be taken into account in the implementation and, especially, tests (or autograd and cudnn implementations won't match).

**Comment by snard6** [comment]
I would love to see this implemented. We have have an old torch7 model we'd like to convert to pytorch that uses CUDNN_SKIP_INPUT. Since it doesn't look like this is coming anytime soon, I'd also be interested in any workarounds to get that model (and its weights) converted.

**Comment by SeanNaren** [comment]
Lots has changed since, closing this for now :)
